watch.go

```diff
@@ -21,53 +20,116 @@ type Event struct {
 	BatchId uint64
 }
 
-// Watcher temporarily stores event information,
+type WatchChan <-chan *Event
+
+type Watcher interface {
+	Watch() WatchChan
+	Close() error
+}
+
+// watcher temporarily stores event information,
 // as it is generated until it is synchronized to DB's watch.
 //
 // If the event is overflow, It will remove the oldest data,
 // even if event hasn't been read yet.
-type Watcher struct {
-	queue eventQueue
-	mu    sync.RWMutex
+type watcher struct {
+	queue   eventQueue
+	mu      sync.RWMutex
+	watchCh chan *Event
+
+	cond     *sync.Cond
+	capacity uint64
+
+	closeMu    sync.Mutex
+	notifyCh   chan struct{}
+	cancelFunc func()
 }
 
-func NewWatcher(capacity uint64) *Watcher {
-	return &Watcher{
+func newWatcher(capacity uint64) *watcher {
+	w := &watcher{
 		queue: eventQueue{
 			Events:   make([]*Event, capacity),
 			Capacity: capacity,
 		},
+		watchCh:  make(chan *Event, 100),
+		cond:     sync.NewCond(&sync.Mutex{}),
+		notifyCh: make(chan struct{}),
+	}
+	w.cancelFunc = func() {
+		close(w.notifyCh)
+		close(w.watchCh)
 	}
+	return w
 }
 
-func (w *Watcher) putEvent(e *Event) {
+func (w *watcher) putEvent(e *Event) {
 	w.mu.Lock()
 	w.queue.push(e)
 	if w.queue.isFull() {
 		w.queue.frontTakeAStep()
 	}
+	if w.capacity < w.queue.Capacity {
+		w.capacity++
+	}
+
+	w.cond.Signal()
 	w.mu.Unlock()
 }
 
 // getEvent if queue is empty, it will return nil.
-func (w *Watcher) getEvent() *Event {
+func (w *watcher) getEvent() *Event {
 	w.mu.RLock()
 	defer w.mu.RUnlock()
 	if w.queue.isEmpty() {
 		return nil
 	}
+	if w.capacity > 0 {
+		w.capacity--
+	}
 	return w.queue.pop()
 }
 
+func (w *watcher) Watch() WatchChan {
+	return w.watchCh
+}
+
+func (w *watcher) Close() error {
+	w.closeMu.Lock()
+	defer w.closeMu.Unlock()
+
+	w.cancelFunc()
+	return nil
+}
+
 // sendEvent send events to DB's watch
-func (w *Watcher) sendEvent(c chan *Event) {
+func (w *watcher) sendEvent() {
 	for {
-		event := w.getEvent()
-		if event == nil {
-			time.Sleep(100 * time.Millisecond)
-			continue
+		select {
+		case <-w.notifyCh:
```

Comment:
Why we need the notifyCh? I do not see any value put into the channel

Reply:
t is only used to notify that the watch has ended. The Close method will close this channel.